src/librustdoc/visit_ast.rs

```diff
@@ -63,23 +63,69 @@ impl<'a, 'tcx> RustdocVisitor<'a, 'tcx> {
     }
 
     pub fn visit(mut self, krate: &'tcx hir::Crate<'_>) -> Module<'tcx> {
-        let mut module = self.visit_mod_contents(
+        let mut top_level_module = self.visit_mod_contents(
             krate.item.span,
             krate.item.attrs,
             &Spanned { span: rustc_span::DUMMY_SP, node: hir::VisibilityKind::Public },
             hir::CRATE_HIR_ID,
             &krate.item.module,
             None,
         );
-        // Attach the crate's exported macros to the top-level module:
-        module
-            .macros
-            .extend(krate.exported_macros.iter().map(|def| self.visit_local_macro(def, None)));
-        module.is_crate = true;
+        top_level_module.is_crate = true;
+        // Attach the crate's exported macros to the top-level module.
+        // In the case of macros 2.0 (`pub macro`), and for built-in `derive`s or attributes as
+        // well (_e.g._, `Copy`), these are wrongly bundled in there too, so we need to fix that by
+        // moving them back to their correct locations.
+        krate.exported_macros.iter().for_each(|def| {
+            /// A return value of `None` signifies a fallback to the default behavior (locating
+            /// the macro at the root of the crate).
+            fn containing_mod_of_macro<'module, 'hir>(
+                def: &'_ rustc_hir::MacroDef<'_>,
+                tcx: TyCtxt<'_>,
+                top_level_module: &'module mut Module<'hir>,
+            ) -> Option<&'module mut Module<'hir>> {
+                // The `def` of a macro in `exported_macros` should correspond to either:
+                //  - a `#[macro-export] macro_rules!` macro,
+                //  - a built-in `derive` (or attribute) macro such as the ones in `::core`,
+                //  - a `pub macro`.
+                // Only the last two need to be fixed, thus:
+                if def.ast.macro_rules {
+                    return None;
+                }
+                /* Because of #77828 we cannot do the simpler:
+                let macro_parent_module = tcx.def_path(tcx.parent_module(def.hir_id).to_def_id());
+                // and instead have to do: */
+                let macro_parent_module = tcx.def_path({
+                    use rustc_middle::ty::DefIdTree;
+                    tcx.parent(tcx.hir().local_def_id(def.hir_id).to_def_id())?
+                });
+                // HACK: rustdoc has no way to lookup `doctree::Module`s by their HirId. Instead,
+                // lookup the module by its name, by looking at each path segment one at a time.
+                // WARNING: this will probably break in the presence of re-exports or shadowing.
```

Comment:
@jyn514 I've been thinking about this: since this whole `visit_ast` happens before `clean`, I don't really see how re-exports could break this (unless the obtained `DefPath` is not what I thought it was (which may be the case 😅)):

 1. IIUC, `rustc` only puts macros in `krate.exported_macros` if they have been defined (and exported) within that very crate. For non-`macro_rules!` macros, their _initial_ location (`DefPath`) is the one traversed through this `for` loop.

 1. The macro or one of its containing modules may be re-exported elsewhere, but that only matters later on, at the `clean` stage.

 1. `clean` will then already to The Right Thing™ in presence of re-exports.

Maybe I should enhance the test file with these weirder scenarios?

```rust
mod inner {
    // ...
    mod private {
        pub macro m() {}
    }

    // @has krate/inner/macro.m.html
    pub use private::m;
}
```

Granted, there is one scenario that does worry me, so I'll add that test case too (I don't know what a `DefPath`s expands to when within a function's scope):

```rust
fn both_fn_and_mod() {
    pub macro m() {}
}
pub mod both_fn_and_mod {
    // @!has krate/both_fn_and_mod/macro.m.html
}
```

Reply:
Hmm, that makes sense to me. Can you add those test cases, and also one for cross-crate re-exports? Search for `aux-build` for examples of how to write a test.

Reply:
Will do!
@rustbot modify labels: -S-waiting-on-review +S-waiting-on-author